neuralpp/symbolic/sympy_expression.py

```diff
@@ -7,41 +7,44 @@
 import builtins
 import fractions
 from neuralpp.symbolic.expression import Expression, FunctionApplication, Variable, Constant, \
-    FunctionNotTypedError, NotTypedError, FunctionType
-from neuralpp.symbolic.basic_expression import BasicConstant, new_type, infer_python_callable_type
-from neuralpp.util.util import consistently_update_dict
+    FunctionNotTypedError, NotTypedError, return_type_after_application, ExpressionType
+from neuralpp.symbolic.basic_expression import infer_python_callable_type
+from neuralpp.util.util import update_consistent_dict
 
 
 # In this file's doc, I try to avoid the term `sympy expression` because it could mean both sympy.Expr (or sympy.Basic)
 # and SymPyExpression. I usually use "sympy object" to refer to the former and "expression" to refer to the latter.
 
 
-def infer_sympy_object_type(sympy_object: sympy.Basic, type_dict: Dict[sympy.Basic, Expression]) -> Expression:
+def infer_sympy_object_type(sympy_object: sympy.Basic, type_dict: Dict[sympy.Basic, ExpressionType]) -> ExpressionType:
+    """
+    type_dict can be, for example, {a: int, b: int, c: float, a+b:int->int->int, (a+b)+c:int->float->float}.
+    """
     match sympy_object:
         case sympy.Integer():
-            return BasicConstant(int)
+            return int
         case sympy.Float():
-            return BasicConstant(float)
+            return float
         case sympy.Rational():
-            return BasicConstant(fractions.Fraction)
+            return fractions.Fraction
         case sympy.logic.boolalg.BooleanAtom():
-            return BasicConstant(bool)
+            return bool
         case _:
             # It is obvious that we can look up type_dict for symbols like `symbol("x")`.
-            # We can also look up for function, such as `a+b`, where type_dict records the function type (instead of
-            # return type) of the (root-level) function.
+            # We can also look up for function application, such as `a+b`, where type_dict records the
+            # function type (instead of return type) of the function used in the function application.
             try:
                 return type_dict[sympy_object]
             except KeyError:  # if it's not in type_dict, try figure out ourselves
-                return BasicConstant(infer_python_callable_type(sympy_function_to_python_callable(sympy_object)))
+                return infer_python_callable_type(sympy_function_to_python_callable(sympy_object))
 
 
-python_callable_and_sympy_function_pair = [
+python_callable_and_sympy_function_relations = [
```

Comment:
Since a relation is a set of tuples (pairs in this case), this should be in the singular.

Reply:
Ah nice catch!